solidity/contracts/libraries/staking/TopUps.sol

```diff
@@ -20,20 +28,71 @@ library TopUps {
+    /// time counter.
+    /// @dev This function should be called only for not yet initialized stake.
+    /// @param value Top-up value, the number of tokens added to the stake.
+    /// @param operator Operator The operator with existing delegation to which
+    /// the tokens should be added to.
+    /// @param operatorParams Parameters of that operator, as stored in the
+    /// staking contract.
+    /// @param escrow Reference to TokenStakingEscrow contract.
+    /// @return New value of parameters. It should be updated for the operator
+    /// in the staking contract.
+    function executeInOneStep(
+        Storage storage self,
+        uint256 value,
+        address operator,
+        uint256 operatorParams,
+        TokenStakingEscrow escrow
+    ) public returns (uint256 newParams) {
+        // Stake is not yet initialized so we don't need to check if the
+        // operator is not undelegating - initializing and undelegating at the
+        // same time is not possible. We do however, need to check whether the
+        // operator has not canceled its previous stake for that operator,
+        // depositing the stake it in the escrow. We do not want to allow
+        // resurrecting operators with cancelled stake by top-ups.
+        require(!escrow.hasDeposit(operator), "Stake canceled");
+
+        uint256 newAmount = operatorParams.getAmount().add(value);
+        newParams = operatorParams.setAmountAndCreationTimestamp(
+            newAmount,
+            block.timestamp
+        );
+
+        emit TopUpCompleted(operator, newAmount);
+    }
+
     /// @notice Initiates top-up of the given value for tokens delegated to
     /// the provided operator. If there is an existing top-up still
     /// initializing, top-up values are summed up and initialization period
     /// is set to the current block timestamp.
+    /// @dev This function should be called only for active operators with
+    /// initialized stake.
     /// @param value Top-up value, the number of tokens added to the stake.
     /// @param operator Operator The operator with existing delegation to which
     /// the tokens should be added to.
+    /// @param operatorParams Parameters of that operator, as stored in the
+    /// staking contract.
+    /// @param escrow Reference to TokenStakingEscrow contract.
     function initiate(
         Storage storage self,
         uint256 value,
-        address operator
+        address operator,
+        uint256 operatorParams,
+        TokenStakingEscrow escrow
     ) public {
+        // Stake is initialized, the operator is still active so we need
+        // to check if it's not undelegating.
+        require(!isUndelegating(operatorParams), "Stake undelegated");
+        // We also need to check if the stake for the operator is not already
+        // in the escrow because it's been previously cancelled.
+        require(!escrow.hasDeposit(operator), "Stake canceled");
```

Comment:
We're not checking if the stake is canceled, but if the operator already deposited anything to the escrow. Maybe we should update the message?
```suggestion
        require(!escrow.hasDeposit(operator), "Operator already deposited to the escrow");
```